crates/full-node/sov-db/src/storage_manager/nomt_based/groups.rs

```diff
@@ -163,101 +168,114 @@ where
+                            continue;
+                        };
+                        let prev_written_version =
+                            user.get_version_for_key(&key.0, previous_version)?;
+
+                        keys_inspected += 1;
+                        if let Some(previous_version) = prev_written_version {
+                            key.1 = previous_version;
+                            batch.delete::<NomtHistoricalState<UserNamespace>>(&key)?;
+                            keys_to_prune += 1;
+                        }
+                        if keys_to_prune >= MAX_INDIVIDUAL_PRUNING_BATCH_SIZE {
+                            hit_size_limit = true;
+                            break;
+                        }
                     }
+                    batch.put::<NomtCommittedVersion<UserNamespace>>(
+                        &VersionedTableMetadataKey::PrunedVersion,
+                        &user_version,
+                    )?;
                 }
-                batch.put::<NomtCommittedVersion<UserNamespace>>(
-                    &VersionedTableMetadataKey::PrunedVersion,
-                    &user_version,
-                )?;
-            }
-            if let Some(kernel_version) =
-                current_kernel_version.and_then(|v| v.checked_sub(versions_to_keep as u64))
-            {
-                let prunable_keys = kernel.iter_pruning_keys_up_to_version(kernel_version)?;
-                for key in prunable_keys {
-                    // Prune the pruning table.
-                    let key = key?;
-                    batch.delete::<NomtPruningState<KernelNamespace>>(&key)?;
-                    keys_to_prune += 1;
-                    keys_inspected += 1;
-                    // Prune the historical state table.
-                    let mut key = key.into_versioned_key();
-                    let Some(previous_version) = key.1.checked_sub(1) else {
-                        continue;
-                    };
-                    let prev_written_version =
-                        kernel.get_version_for_key(&key.0, previous_version)?;
-                    keys_inspected += 1;
-                    if let Some(previous_version) = prev_written_version {
-                        key.1 = previous_version;
-                        batch.delete::<NomtHistoricalState<KernelNamespace>>(&key)?;
+                if let Some(kernel_version) =
+                    current_kernel_version.and_then(|v| v.checked_sub(versions_to_keep as u64))
+                {
+                    let prunable_keys = kernel
+                        .iter_pruning_keys_up_to_version(kernel_version)?
+                        .take(MAX_INDIVIDUAL_PRUNING_BATCH_SIZE);
+                    for key in prunable_keys {
+                        // Prune the pruning table.
+                        let key = key?;
+                        batch.delete::<NomtPruningState<KernelNamespace>>(&key)?;
```

Comment:
This for loop looks very similar to the previous one. Let's extract the core logic to a helper method.

Reply:
Going to leave this out of scope since it's not a new issue. Happy to handle in a follow-up, but let's get this fix merged cuz it could break zeta.